### ISSUE TITLE
Training with a dataset that contains different sized images. I want to use batch size 1 without img resize. 

### ISSUE THREAD

**Issue by FurkanYlmz97** [ultralytics/yolov5#1497]
## ❔Question

All right I can train the Yolo as following,
"$ python train.py --img 640 --batch 16 --epochs 5 --data coco128.yaml --weights yolov5s.pt"

As far as I understood to train as a batch we give the image size as input (default: [640, 640]). However, doesn't changing the size of the picture reduce performance a little bit? I think in my work not changing the input image size is important. Therefore, I want to train on a dataset with batch size 1 without changing the input image's size. Are there any options or how can I trick the code to train like that?

**Comment by glenn-jocher** [comment]
@FurkanYlmz97 training at batch-size 1 is not recommended, as the BatchNorm2d() layers will not acquire sufficient statistics in order to properly normalize full batches, i.e. your mAP will suffer all else equal.

Recommend you simply train with all default settings, or at the --img-size you expect to run inference at. See https://docs.ultralytics.com/yolov5/tutorials/train_custom_data

**Comment by pfeatherstone** [comment]
What if you're not really training on images but 2D blobs and bilinear resizing makes very little sense and wouldn't be used during inference anyway ?

**Comment by pfeatherstone** [comment]
@glenn-jocher have you experimented with training at different sizes? Like darknet will train at 416x416 for a bit, then train at 512x512 and so on. Have you tried doing something similar ?

**Comment by pfeatherstone** [comment]
For example, you might want to do object detection with audio. To do that, You might want to convert audio to 2D blobs using spectrograms. But resizing a spectrogram using bilinear resizing makes no sense. If you want a different shape, you change the FFT size, the window length, the hop length and the number of averages per trace. So very different computation to cv2.resize(). So you might build a dataset of audio spectrograms of different shapes. How can I train on those without simply doing an image resize to something standard like 640x640.

**Comment by glenn-jocher** [comment]
@pfeatherstone you can train at different image sizes with --multi-scale.

YOLOv5 trains with scale jitter automatically already:
https://github.com/ultralytics/yolov5/blob/4c5d9bff80526b1120b2f78ce81bd20ec1a50b4e/data/hyp.scratch.yaml#L27

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/122036784-3cb22200-cdd4-11eb-9adf-b123277cd7dd.jpg)

**Comment by pfeatherstone** [comment]
Cool thank you. I think that still uses bilinear resizing. My scenario is I have datasets, each containing 2D blobs of size [400,600], [1000,256], [128,4096], ... for example. So each dataset has a specific size. Then i want to concatenate them without doing any resizing. So when doing batch loading, i need to make sure each batch contains elements from one of the datasets. Then what is the best strategy: randomly alternate the batches between the datasets, or, train on dataset A for a few epochs, containing blobs of size [400,600] say, then train for a few epochs on dataset B, containing blobs of size [1000,256] for example. What i don't want to do is naively concatenate the datasets, then when batchloading everything, use cv2.resize or F.interpolate() to normalize to a common size, then do what you do with multi-scale and randomly resize the batch. Coz, using F.interploate(.., mode='bilinear') doesn't make sense for things like spectrograms. If you want to resize a spectrogram, you go back to the beginning of the computation, choose an appropriate FFT size and window length.

**Comment by pfeatherstone** [comment]
The main question is, what is the best strategy when training on different sizes? I noticed you simply randomize the size of each batch when multi-scale is enabled. When i've tried that, training is very unstable. Maybe that's why darknet picks a size, trains for a few epochs, then chooses another size, trains for a few epochs, then continue. But that seems wrong to me. so maybe there is a way to stabilize training when randomly resizing every batch. Maybe there is a way to scale the gradients somehow...

**Comment by pfeatherstone** [comment]
Maybe it's because there isn't a good set of anchors that fits all datasets. So no matter what you do, the gradients are high at the yolo layers, which makes the whole thing unstable.

**Comment by glenn-jocher** [comment]
@pfeatherstone --multi-scale does not cause instabilities, if you'd like to train at various size this is the correct way.

**Comment by pfeatherstone** [comment]
Unless using something like F.interpolate() is not possible (at the input)

**Comment by pfeatherstone** [comment]
I think the short answer is: this can't really be used as-is for audio processing.
